packages/detectors.py

```diff
@@ -0,0 +1,256 @@
+    vol_shape = data.shape[:-1]
+    #- 'n_vols' is the number of volumes
+    n_vols = data.shape[-1]
+    #- N is the number of voxels in a volume
+    N = np.prod(vol_shape)
+
+    #- Reshape to 2D array that is voxels by volumes (N x n_vols)
+    # transpose to n_vols x N
+    X = data.reshape((N, n_vols)).T
+
+    """
+    The first part of the code will use PCA to get component matrix U
+    and scalar projections matrix C
+    """
+
+    #- Calculate unscaled covariance matrix for X
+    unscaled_cov = X.dot(X.T)
+
+    #- Use SVD to return U, S, VT matrices from unscaled covariance
+    U, S, VT = npl.svd(unscaled_cov)
+
+    #- Calculate the scalar projections for projecting X onto the vectors in U.
+    #- Put the result into a new array C.
+    C = U.T.dot(X)
+    # set nans to 0
+    C[np.isnan(C)] = 0
+    #- Transpose C
+    #- Reshape C to have the 4D shape of the original data volumes.
+    C_vols = C.T.reshape((vol_shape + (n_vols,)))
+
+    """
+    The second part of the code determines which voxels are inside the brain
+    and which are outside the brain and creates a mask (boolean matrix)
+    """
+
+    #get the mean voxel intensity of entire 4D object
+    mean_voxel = np.mean(data)
+    #get the mean volume (3D) across time series (axis 3)
+    mean_volume = np.mean(data, axis=3)
+    #boolean mask set to all voxels above .5 in the first volume
+    #(.125 is the SPM criterion but .5 seems like a better threshold)
+    mask = mean_volume > (.5 * mean_voxel) #threshold can be adjusted!
+    out_mask = ~mask
+
+    """
+    The third part of code finds the root mean square of U from step 1, then uses the
+    mask from step 2 to determine which components explain data outside the brain
+    Selects these "bad components" with high "outsideness"
+    """
+
+    #Apply mask to C matrix to get all voxels outside of brain
+    outside = C_vols[out_mask]
+    #Get RMS of the voxels outside, reflecting "outsideness" of this scan
+    RMS_out = np.sqrt(np.mean((outside ** 2), axis=0))
+
+    #Apply mask to C matrix to get all voxels inside brain
+    inside = C_vols[mask]
+    #Get RMS of the voxels inside, reflecting "insideness" of this scan
+    RMS_in = np.sqrt(np.mean((inside ** 2), axis=0))
+
+    #The closer this ratio is to 1, the worse the volume
+    RMS_ratio = RMS_out / RMS_in
+
+    """
+    The fourth part of the code uses the "bad components" to generate a new
+    "bad data set" and then puts this dataset through the outlier detector
+    """
+
+    #Create a boolean mask for the 10% worst PCs (meaning highest RMS ratio)
+    PC_bad = np.percentile(RMS_ratio, 90)
+    PC_bad_mask = RMS_ratio > PC_bad
+
+    U_bad = U[:, PC_bad_mask]
+    C_bad = C[PC_bad_mask]
+
+    #generates data set based on the bad PCs and (U and C matrices)
+    X_bad = U_bad.dot(C_bad).T.reshape((vol_shape + (n_vols,)))
+
+    # calculate outliers using iqr_detector
+    _, outliers = mah_detector(X_bad)
```

Comment:
I think the better code design here would be to return the vector of measures from the function (here - "badness") and then call the generic detector functions such as `mah_detector` on the vectors.  That allows you to mix and match the measures and detectors more easily, and makes your code more re-usable.